File: 01-observer-pattern/01-observer-clock-example-in-java/src/example/Subject.java
package example;

import java.util.ArrayList;
import java.util.List;

public class Subject<T> {
    private final List<Observer<T>> observers = new ArrayList<>();

    public void register(Observer<T> observer) {
        this.observers.add(observer);
    }

    protected void notifyObservers(T value) {
        this.observers.forEach(o -> o.update(value));
    }
}

File: 01-observer-pattern/02-weather-station-refactor/src/subject.ts
import {WeatherRecording} from "src/weather-recording.ts";

export class Subject {
    private readonly observers: Observer[] = [];

    public register(observer: Observer): void {
        this.observers.push(observer);
    }

    protected notify(recording: WeatherRecording) {
        this.observers.forEach(o => o.update(recording))
    }
}

export interface Observer {
    update(recording: WeatherRecording): void;
}
